envoy/pkg/livenessprobe/cmd/root.go

```diff
@@ -0,0 +1,168 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/well"
+	"github.com/spf13/cobra"
+)
+
+var config struct {
+	timeout    time.Duration
+	listenAddr string
+	readyURL   string
+	httpURL    string
+	httpsAddr  string
+}
+
+var rootCmd = &cobra.Command{
+	Use:   "livenessprobe",
+	Short: "liveness probe for Envoy",
+	Long:  `Liveness probe for Envoy.`,
+
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		cmd.SilenceUsage = true
+
+		err := well.LogConfig{}.Apply()
+		if err != nil {
+			log.ErrorExit(err)
+		}
+	},
+
+	RunE: func(cmd *cobra.Command, args []string) error {
+		mux := http.NewServeMux()
+		mux.HandleFunc("/", handler)
+		serv := &well.HTTPServer{
+			Server: &http.Server{
+				Addr:    config.listenAddr,
+				Handler: mux,
+			},
+		}
+		err := serv.ListenAndServe()
+		if err != nil {
+			return err
+		}
+
+		err = well.Wait()
+		if err != nil && !well.IsSignaled(err) {
+			return err
+		}
+		return nil
+	},
+}
+
+// Execute executes the command.
+func Execute() {
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func handler(rw http.ResponseWriter, req *http.Request) {
+	env := well.NewEnvironment(req.Context())
+	env.Go(monitorReady)
+	env.Go(monitorHTTP)
+	env.Go(monitorHTTPS)
+
+	env.Stop()
+	err := env.Wait()
+	if err != nil {
+		rw.WriteHeader(http.StatusBadGateway)
+		rw.Write([]byte(err.Error()))
+		return
+	}
+
+	rw.WriteHeader(http.StatusOK)
+	rw.Write([]byte("ok"))
+	return
+}
+
+func monitorReady(ctx context.Context) error {
+	c := well.HTTPClient{
```

Comment:
Do not throw away http.Client.  The package is not designed so.
https://golang.org/pkg/net/http/#Client

> The Client's Transport typically has internal state (cached TCP connections), so **Clients should be reused** instead of created as needed. Clients are safe for concurrent use by multiple goroutines.

Also, use a private http.Transport and configure it not to keep-alive connections.
```go
    tp := &http.Transport{
        DisableKeepAlives: true,

        // rest are copied from http.DefaultTransport
        Proxy: ProxyFromEnvironment,
        DialContext: (&net.Dialer{
            Timeout:   30 * time.Second,
            KeepAlive: 30 * time.Second,
            DualStack: true,
        }).DialContext,
        ForceAttemptHTTP2:     true,
        MaxIdleConns:          100,
        IdleConnTimeout:       90 * time.Second,
        TLSHandshakeTimeout:   10 * time.Second,
        ExpectContinueTimeout: 1 * time.Second,
    }
```